File: sdk/a2a.py
import httpx
from typing import Optional


class A2AClient:
    def __init__(self, server_url: str, timeout: float = 30.0):
        self._base = server_url.rstrip("/")
        self._client = httpx.AsyncClient(timeout=timeout)

    async def initiate(
        self,
        sender_did: str,
        receiver_did: str,
        amount: int,
        description: str,
        service_type: Optional[str] = None,
        timeout_minutes: int = 60,
        currency: str = "jpy",
    ) -> dict:
        resp = await self._client.post(
            f"{self._base}/a2a/transfer",
            json={
                "sender_did": sender_did,
                "receiver_did": receiver_did,
                "amount": amount,
                "currency": currency,
                "description": description,
                "service_type": service_type,
                "timeout_minutes": timeout_minutes,
            },
        )
        resp.raise_for_status()
        return resp.json()

    async def get_status(self, transfer_id: str) -> dict:
        resp = await self._client.get(f"{self._base}/a2a/transfer/{transfer_id}")
        resp.raise_for_status()
        return resp.json()

    async def complete(self, saga_id: str, reporter_did: str, result_summary: str = "") -> dict:
        resp = await self._client.post(
            f"{self._base}/saga/{saga_id}/complete",
            json={"reporter_did": reporter_did, "result_summary": result_summary},
        )
        resp.raise_for_status()
        return resp.json()

    async def close(self):
        await self._client.aclose()

File: sdk-ts/src/a2a.ts
export interface A2ATransferRequest {
  senderDid: string;
  receiverDid: string;
  amount: number;
  description: string;
  serviceType?: string;
  timeoutMinutes?: number;
  currency?: string;
}

export interface A2ATransferResponse {
  transfer_id: string;
  saga_id: string;
  status: string;
  escrow_status: string;
  timeout_at: string | null;
  steps: { total: number; completed: number; current: string };
}

export class A2AClient {
  constructor(private readonly serverUrl: string) {}

  async initiate(req: A2ATransferRequest): Promise<A2ATransferResponse> {
    const res = await fetch(`${this.serverUrl}/a2a/transfer`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        sender_did: req.senderDid,
        receiver_did: req.receiverDid,
        amount: req.amount,
        currency: req.currency ?? 'jpy',
        description: req.description,
        service_type: req.serviceType,
        timeout_minutes: req.timeoutMinutes ?? 60,
      }),
    });
    if (!res.ok) throw new Error(`A2A initiate failed: ${await res.text()}`);
    return res.json();
  }

  async getStatus(transferId: string): Promise<Record<string, unknown>> {
    const res = await fetch(`${this.serverUrl}/a2a/transfer/${transferId}`);
    if (!res.ok) throw new Error(`Get transfer failed: ${await res.text()}`);
    return res.json();
  }

  async complete(sagaId: string, reporterDid: string, resultSummary = ''): Promise<Record<string, unknown>> {
    const res = await fetch(`${this.serverUrl}/saga/${sagaId}/complete`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ reporter_did: reporterDid, result_summary: resultSummary }),
    });
    if (!res.ok) throw new Error(`Complete saga failed: ${await res.text()}`);
    return res.json();
  }
}
